plugins/BEdita/Core/src/Model/Table/TreesTable.php

```diff
@@ -99,9 +104,76 @@ public function validationDefault(Validator $validator)
     public function buildRules(RulesChecker $rules)
     {
         $rules->add($rules->existsIn(['object_id'], 'Objects'));
-        $rules->add($rules->existsIn(['parent_id'], 'ParentTrees'));
-        $rules->add($rules->existsIn(['root_id'], 'Objects'));
+        $rules->add($rules->existsIn(['root_id'], 'RootObjects'));
+        $rules->add($rules->existsIn(
+            ['parent_id'],
+            'ParentObjects',
+            ['allowNullableNulls' => true]
+        ));
+        $rules->add($rules->existsIn(
+            ['parent_node_id'],
+            'ParentNode',
+            ['allowNullableNulls' => true]
+        ));
+
+        $rules->add(
+            [$this, 'isParentValid'],
+            'isParentValid',
+            [
+                'errorField' => 'parent_id',
+                'message' => __d('bedita', 'parent_id must be null or corresponding to a folder'),
+            ]
+        );
 
         return $rules;
     }
+
+    /**
+     * Check that `parent_id` property of the entity corresponds to a folder
+     *
+     * @param EntityInterface $entity The entity to validate
+     * @return bool
+     */
+    public function isParentValid(EntityInterface $entity)
+    {
+        if ($entity->parent_id === null) {
+            return true;
```

Comment:
`parent_id` can be `null` if and only if `$entity->object_id` is an ID of a folder. Is it worth checking?

Reply:
:thinking: Without that check the following `find()` (that will become `exists()`) see https://github.com/bedita/bedita/pull/1408#discussion_r175834784) will return `false`, but a `parent_id === null` is valid value for root folder.

Reply:
No, I'd leave this check, but instead of returning `true` straight away, we should probably do something like:
```php
if ($entity->parent_id === null) {
    return $Objects->exists(['id' => $entity->object_id, 'object_type_id' => $foldersOTID]);
}
```

Or is this check only performed for folder nodes?

Reply:
Ahhhh ok, I understand. It's not only for folder node, is used as applcation rule for every node so you are right :+1: